### DISCUSSION TITLE
Limit bounding boxes

### DISCUSSION THREAD

**Discussion by amandakj** [ultralytics/yolov5#2248] [Q&A] [answered]
Hi.

I have images with three different objects, where they only appear once. When I am training and detecting I sometimes get multiple bounding boxes for each class. I want to limit the number of bounding boxes to one per class. 

Is there a hyperparameter or a function I can change to easily fix this problem?

**Comment by glenn-jocher** [comment, marked as answer]
@amandakj yes, you can control this by varying NMS settings to suit your needs:

- **Agnostic NMS** will run NMS across all classes at once, eliminating incidences of overlapping boxes from different classes (False by default). To enable: `python detect.py --agnostic`
- **Multilabel NMS** assigns all classes above threshold to the same bounding box. Enabled by default. You can disable here: https://github.com/ultralytics/yolov5/blob/9d873077841434d1c6cbd1c4248ca2252820d3ba/utils/general.py#L409

**Reply by amandakj** [reply]
Thanks, @glenn-jocher! 😄

**Reply by alucor-it** [reply]
Hi @glenn-jocher ,
I have set --agnostic-nms but I still get multiple bounding boxes per class in my saved text file. Do I have to make any other changes?

**Reply by glenn-jocher** [reply]
@alucor-it multiple detections within one class in an image is not incompatible with agnostic NMS. See https://www.quora.com/What-does-%E2%80%9Cclass-agnostic%E2%80%9D-in-most-of-the-object-detection-papers-mean

**Comment by glenn-jocher** [comment]
@amandakj we are modifying the default inference behavior here to `multi_label=False` here after multiple user feedback. See https://github.com/ultralytics/yolov5/pull/2252 for details.

**Comment by jeannotes** [comment]
1. I think the multilabel only reference to those different objects but close to each other and overlap closely,  right?
2. but when enable multi-label, what is the insight, meaningful?

**Reply by glenn-jocher** [reply]
@jeannotes multi-label means one bounding box can have multiple classes.

**Reply by jeannotes** [reply]
why, what's this for?

**Reply by glenn-jocher** [reply]
@jeannotes this is for use cases that require this configuration.

**Reply by jeannotes** [reply]
@glenn-jocher why needs this requirements?

**Reply by glenn-jocher** [reply]
@jeannotes it's not a requirement.

**Reply by SpongeBab** [reply]
@jeannotes ,it not must, it just need or not.

**Reply by jeannotes** [reply]
@glenn-jocher @SpongeBab  but I still don't understand that when set "multi_label = nc > 1 ", what is the advantages

**Comment by myasser63** [comment]
@glenn-jocher How can I set yolov5 for multiclass not multilabel task? is this done by Agnostic NMS only or there's another setting?

**Reply by myasser63** [reply]
@glenn-jocher Do I need to change Cls Loss from BCEwithlogits to CrossEntropy loss to convert to MultiClass ?

**Reply by glenn-jocher** [reply]
@myasser63 multi-label NMS will allow for multiple classes on a single box:
https://github.com/ultralytics/yolov5/blob/d7bc5d77730c625e8b7039e8680714602f5b321f/utils/general.py#L806-L813

**Reply by myasser63** [reply]
@glenn-jocher Sorry I don't fully understand. My dataset is having 4 classes mutually execlusive. From my understanding `BCEwithlogits` is for Mulitlabel problems. So to make only the task for multiclass only, just make `multi_label` to `False` without changing the loss function?. I mean during training and validation not inference.

**Reply by glenn-jocher** [reply]
@myasser63 yes BCE is multi-label capable, but by default multi-label is False and NMS is set to output only 1 label per box.
